spec.html

```diff
@@ -37244,76 +37244,96 @@ <h1>ValidateAtomicAccess ( _typedArray_, _requestIndex_ )</h1>
+        <p>A <dfn id="waiterlist-record">WaiterList Record</dfn> is a Record value used to explain waiting and notification of agents via `Atomics.wait` and `Atomics.notify`. It has fields as defined by <emu-xref href="#table-waiterlist-record-fields"></emu-xref>.</p>
+        <emu-table id="table-waiterlist-record-fields" caption="WaiterList Record Fields">
+          <table>
+            <tbody>
+            <tr>
+              <th>Field Name</th>
+              <th>Value</th>
+              <th>Meaning</th>
+            </tr>
+            <tr>
+              <td>[[Waiters]]</td>
+              <td>A list of <emu-xref href="#agent">agents</emu-xref>.</td>
+              <td>The list of agents waiting on a memory location.</td>
+            </tr>
+            <tr>
+              <td>[[MostRecentLeaveEvent]]</td>
+              <td>Synchronize event | *undefined*</td>
+              <td>The event of the most recent leaving of its critical section, or *undefined* if its critical section has never been entered. </td>
+            </tr>
+            </tbody>
+          </table>
+        </emu-table>
+        <p>The agent cluster has a store of WaiterList Records; the store is indexed by (_block_, _i_). WaiterList Records are agent-independent: a lookup in the store of WaiterList Records by (_block_, _i_) will result in the same WaiterList Record in any agent in the agent cluster.</p>
+        <p>Each WaiterList Record has a <dfn>critical section</dfn> that controls exclusive access to that WaiterList Record during evaluation. Only a single agent may enter a WaiterList Record's critical section at one time. Entering and leaving a WaiterList Record's critical section is controlled by the abstract operations EnterCriticalSection and LeaveCriticalSection. Operations on a WaiterList Record&mdash;adding and removing waiting agents, traversing the list of agents, suspending and notifying agents on the list, setting and retrieving the Synchronize event&mdash;may only be performed by agents that have entered the WaiterList Record's critical section.</p>
         <p>The abstract operation GetWaiterList takes arguments _block_ (a Shared Data Block) and _i_ (a non-negative integer). It performs the following steps when called:</p>
         <emu-alg>
           1. Assert: _block_ is a Shared Data Block.
           1. Assert: _i_ and _i_ + 3 are valid byte offsets within the memory of _block_.
           1. Assert: _i_ is divisible by 4.
-          1. Return the WaiterList that is referenced by the pair (_block_, _i_).
+          1. Return the WaiterList Record that is referenced by the pair (_block_, _i_).
```

Comment:
This part is a tricky. The "nice" thing about WaiterList being a handwavy magical "semantic object" was that we sidestepped making synchronization around this agent cluster-wide WaiterList store explicit. We can kind of pretend that they magically appear, or that they always exist for any possible (_block_, _i_).

Now, since for Records, the creation of new Records is an explicit thing, we now need to also explicitly capture the creation of these WaiterList records in the agent-wide store. Importantly, we need to synchronize the creation. If two agents are concurrently trying to get a WL corresponding to (_b_, _i_), we need to ensure both agents get the same WL.

To capture this, we need something like the following:

- The agent cluster-wide WL store itself needs a critical section that GetWaiterList enters.
- If the store doesn't have a record for (_b_, _i_), one needs to be created and explicitly set [[Waiters]] to the empty List and [[MostRecentLeaveEvent]] to ~empty~.

That's more machinery for not a lot of extra clarity of understanding, so I'm on the fence for this PR.